ddl/column.go

```diff
@@ -222,14 +222,252 @@ func onAddColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error)
 
 		// Finish this job.
 		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
-		asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfo: columnInfo})
+		asyncNotifyEvent(d, &util.Event{Tp: model.ActionAddColumn, TableInfo: tblInfo, ColumnInfos: []*model.ColumnInfo{columnInfo}})
 	default:
 		err = ErrInvalidDDLState.GenWithStackByArgs("column", columnInfo.State)
 	}
 
 	return ver, errors.Trace(err)
 }
 
+func checkAddAndCreateColumnInfos(t *meta.Meta, job *model.Job) (*model.TableInfo, []*model.ColumnInfo, []*ast.ColumnPosition, []int, error) {
+	schemaID := job.SchemaID
+	tblInfo, err := getTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	columns := []*model.ColumnInfo{}
+	positions := []*ast.ColumnPosition{}
+	offsets := []int{}
+	err = job.DecodeArgs(&columns, &positions, &offsets)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, nil, nil, nil, errors.Trace(err)
+	}
+	// Should initialize positions[i] when it is nil, otherwise the createColumnInfo function will panic.
+	for i := range positions {
+		if positions[i] == nil {
+			positions[i] = &ast.ColumnPosition{}
+		}
+	}
+
+	var columnInfos []*model.ColumnInfo
+
+	for i, col := range columns {
+		columnInfo := model.FindColumnInfo(tblInfo.Columns, col.Name.L)
+		if columnInfo != nil {
+			columnInfos = append(columnInfos, columnInfo)
+			if columnInfo.State == model.StatePublic {
+				// We already have a column with the same column name.
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, nil, infoschema.ErrColumnExists.GenWithStackByArgs(col.Name)
+			}
+		} else {
+			columnInfo, offset, err := createColumnInfo(tblInfo, col, positions[i])
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return nil, nil, nil, nil, errors.Trace(err)
+			}
+			logutil.BgLogger().Info("[ddl] run add columns job", zap.String("job", job.String()), zap.Reflect("columnInfo", *columnInfo), zap.Int("offset", offsets[i]))
+
+			// Set offset arg to job.
+			if offset != 0 {
```

Comment:
Could we remove this statement of `if`? If `position.Tp` is `ColumnPositionFirst`, the offset is 0.